cpu_scheduling_algorithms/round_robin_scheduling.cpp

```diff
@@ -0,0 +1,248 @@
+};
+
+/**
+ * Represent remaining burst time of a process.
+ */
+using BTLeft = uint32_t;
+
+/**
+ * @brief Execute processes based on Round-robin algorithm.
+ *
+ * \param processes Processes to be executed
+ * \param time_slice Time slice for processes execution
+ * \return Results of each process execution
+ */
+std::vector<ProcessResult> RRExecute(const std::vector<Process>& processes,
+                                     uint32_t time_slice);
+
+/**
+ * @brief Print a table containing process results data.
+ *
+ * \return ostream inputted ostream
+ */
+std::ostream& operator<<(std::ostream& ostream,
+                         const std::vector<ProcessResult>& results);
+
+/**
+ * @brief Self-test implementations.
+ *
+ * \returns void
+ */
+static void Test();
+
+/**
+ * @brief Comparator function for sorting processes.
+ *
+ * \param p1 Process to be compared
+ * \param p2 Process to be compared
+ * \return
+ */
+bool CompareAT(const Process& p1, const Process& p2) {
+    return p1.arrival_time < p2.arrival_time;
+}
+
+/**
+ * @brief Check processes that arrive after the given time_elapsed.
+ *
+ * If a process arrive after the give time_elapsed, then the process
+ * will be pushed into the schedule queue and inserted into the
+ * arrived_process set.
+ *
+ * \param processes Processes that will be checked for arrival
+ * \param arrived_process A set containing processes that has arrived
+ * \param schedule Queue containing pair of process and its remaining burst time
+ * \param time_elapsed Time that has elapsed after processes execution
+ */
+void CheckArriveProcess(const std::vector<Process>& processes,
+                        std::set<uint32_t>* arrived_process,
+                        std::queue<std::pair<Process, BTLeft>>* schedule,
+                        uint32_t time_elapsed);
+
+/**
+ * @brief Entry point of the program.
+ *
+ * \return 0 on exit
+ */
+int main() {
+    Test();
+    return 0;
+}
+
+/**
+ * @brief Self-test implementations
+ * @returns void
+ */
+static void Test() {
+    std::vector<Process> processes{
+        {0, 70, 3}, {1, 9, 2}, {2, 3, 39}, {3, 5, 29}, {4, 30, 90}};
+    const uint32_t kTimeSlice{3};
+    std::vector<ProcessResult> results = RRExecute(processes, kTimeSlice);
+
+    std::vector<uint32_t> correct_completion_times({80, 14, 100, 82, 166});
+    std::vector<ProcessResult> correct_results;
+    // Generate correct process results based on correct completion times
+    for (size_t i = 0; i < processes.size(); i++) {
+        correct_results.emplace_back(processes[i], correct_completion_times[i]);
+    }
+
+    // Sort the results and correct results so they're exactly the same
+    std::sort(results.begin(), results.end(), CompareAT);
+    std::sort(correct_results.begin(), correct_results.end(), CompareAT);
+
+    std::cout << results;
+    assert(results == correct_results);
+    std::cout << "All test passed";
+}
+
+std::vector<ProcessResult> RRExecute(const std::vector<Process>& processes,
```

Comment:
Functions should be documented explaining the use of its parameters, a brief description, and what it returns. Check other functions as well.

Reply:
I have added functions documentation at their function prototype. Should I add documentation to their function definition too?